eth/vm/computation.py

```diff
@@ -559,30 +577,37 @@ def apply_create_message(self) -> 'BaseComputation':
     def apply_computation(cls,
                           state: BaseState,
                           message: Message,
-                          transaction_context: BaseTransactionContext) -> 'BaseComputation':
+                          transaction_context: BaseTransactionContext,
+                          tracer: BaseTracer) -> 'BaseComputation':
         """
         Perform the computation that would be triggered by the VM message.
         """
-        with cls(state, message, transaction_context) as computation:
+        with cls(state, message, transaction_context, tracer) as computation:
             # Early exit on pre-compiles
             if message.code_address in computation.precompiles:
                 computation.precompiles[message.code_address](computation)
                 return computation
 
             for opcode in computation.code:
                 opcode_fn = computation.get_opcode_fn(opcode)
+                pc = computation.get_pc()
 
                 computation.logger.debug2(
                     "OPCODE: 0x%x (%s) | pc: %s",
                     opcode,
                     opcode_fn.mnemonic,
-                    max(0, computation.code.pc - 1),
+                    pc,
                 )
 
                 try:
-                    opcode_fn(computation=computation)
+                    with computation.tracer.capture(computation, opcode_fn):
+                        opcode_fn(computation=computation)
                 except Halt:
                     break
+
+        # Allow the tracer to record final values.
+        computation.tracer.finalize(computation)
```

Comment:
I'm exhausted my timebox but I was planning at looking at this next. I think this fails if there are any child computations? When the child's apply_computation finishes the tracer is finalized. It then returns to the parent which attempts to continue tracing and the tracer will complain that it's already been finalized.